File: backend/app/models/recurring_item.py
from datetime import UTC, datetime
from decimal import Decimal
from enum import StrEnum

from beanie import Indexed
from pydantic import Field

from app.models.base import AyeDecimal, SoftDeleteDocument


class RecurringType(StrEnum):
    ingreso_fijo = "ingreso_fijo"
    gasto_fijo = "gasto_fijo"
    mensualidad = "mensualidad"


class Frequency(StrEnum):
    semanal = "semanal"
    quincenal = "quincenal"
    mensual = "mensual"


class RecurringItem(SoftDeleteDocument):
    user_id: Indexed(str)
    name: str
    type: RecurringType
    amount: AyeDecimal = Field(default=Decimal("0.00"))
    frequency: Frequency = Frequency.mensual
    day_of_month: int | None = Field(default=None, ge=1, le=31)
    account_id: Indexed(str)
    next_date: datetime = Field(default_factory=lambda: datetime.now(UTC))
    is_active: bool = True

    class Settings:
        name = "recurring_items"
        use_state_management = True

File: ios-app/AyeFinance/Models/RecurringItem.swift
import Foundation

public enum RecurringType: String, Codable, CaseIterable {
    case ingresoFijo = "ingreso_fijo"
    case gastoFijo = "gasto_fijo"
    case mensualidad = "mensualidad"
    
    public var displayName: String {
        switch self {
        case .ingresoFijo: return "Ingreso Fijo"
        case .gastoFijo: return "Gasto Fijo"
        case .mensualidad: return "Mensualidad"
        }
    }
}

public enum Frequency: String, Codable, CaseIterable {
    case semanal
    case quincenal
    case mensual
    
    public var displayName: String {
        switch self {
        case .semanal: return "Semanal"
        case .quincenal: return "Quincenal"
        case .mensual: return "Mensual"
        }
    }
}

public struct RecurringItem: Codable, Identifiable, Hashable {
    public let id: String
    public let userId: String
    public var name: String
    public var type: RecurringType
    public var amount: String
    public var frequency: Frequency
    public var dayOfMonth: Int?
    public var accountId: String
    public var nextDate: String
    public var isActive: Bool
    public let createdAt: String
    public let updatedAt: String
    
    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case name
        case type
        case amount
        case frequency
        case dayOfMonth = "day_of_month"
        case accountId = "account_id"
        case nextDate = "next_date"
        case isActive = "is_active"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
    
    public var amountDouble: Double {
        Double(amount) ?? 0.0
    }
}
